### ISSUE TITLE
Wesnoth crashes on Windows: MinGW Runtime Assertion failed

### ISSUE THREAD

**Issue by cooljeanius** [wesnoth/wesnoth#6030]
I'm trying to introduce my fiancée to Wesnoth, but unfortunately crashes are preventing it from going too well; the error dialog box says it's the assertion at this line for `dst` that's failing: https://github.com/wesnoth/wesnoth/blob/8d6c8ef78de0e4b78ef1a929b80eafae99d2d9aa/src/sdl/utils.cpp#L2013
This is Wesnoth 1.14.17, installed via Steam on Windows 10 Home. It's possibly another manifestation of the whole OneDrive issue, but not one I've seen mentioned in any other discussions of OneDrive issues, so I'm not sure.

**Comment by irydacea** [comment]
Can you provide steps to reproduce? The assertion check alone doesn't say anything about the issue.

**Comment by cooljeanius** [comment]
wait nvm I think this is a dup of #4855; I forgot to include closed issues when searching previously...

**Comment by cooljeanius** [comment]
last bit of log file before crash looks like this, btw:
```
20210816 16:38:00 error deprecation: MENU_IMG_TXT2 has been deprecated and will be removed in version 1.15.0.
  Use the new [difficulty] tag instead for campaign difficulties, or the label/description keys for message options.
Failed to create a surface of size 1738x1050 Reason: Out of memory
Could not create a new surface in get_surface_portion()
Failed to create a surface of size 1738x1050 Reason: Out of memory
20210816 16:42:27 error gui/draw: SDL_CreateSoftwareRenderer failed, reason: Can't create renderer for NULL surface
Failed to create a surface of size 1738x1050 Reason: Out of memory
20210816 16:42:27 error gui/draw: SDL_CreateSoftwareRenderer failed, reason: Can't create renderer for NULL surface
```

**Comment by cooljeanius** [comment]
> wait nvm I think this is a dup of #4855; I forgot to include closed issues when searching previously... 

Although, wait, #4855 was closed as a dup of #4934, but 4934 was closed as a dup of 4855... how do circular duplicate cycles like that work?

**Comment by Pentarctagon** [comment]
#4934 was closed because #4855 predates it.   #4855 was closed because the issue was fixed, and that issue is the same issue that was reported in #4934.

**Comment by Pentarctagon** [comment]
@cooljeanius 
> last bit of log file before crash looks like this, btw:

Right, but what was being done prior to the crash?  The bug that caused both those previous reports was fixed in 1.14.13.   Also, the information listed [here](https://forums.wesnoth.org/viewtopic.php?f=4&t=25660) (as applicable) is useful.

**Comment by cooljeanius** [comment]
info from `i` button:
```
The Battle for Wesnoth version 1.14.17
Running on Microsoft Windows 10 (10.0.19043)
Distribution channel: Steam

Game paths
==========

Data dir:        D:\Steam\steamapps\common\wesnoth
User config dir: C:\Users\USER\OneDrive\Documents\My Games\Wesnoth1.14
User data dir:   C:\Users\USER\OneDrive\Documents\My Games\Wesnoth1.14
Saves dir:       C:\Users\USER\OneDrive\Documents\My Games\Wesnoth1.14\saves
Add-ons dir:     C:\Users\USER\OneDrive\Documents\My Games\Wesnoth1.14\data\add-ons
Cache dir:       C:\Users\USER\OneDrive\Documents\My Games\Wesnoth1.14\cache

Libraries
=========

Boost:             1.66
OpenSSL/libcrypto: 1.1.0f (runtime 1.1.0f)
Cairo:             1.10.2 (runtime 1.10.2)
Pango:             1.30.1 (runtime 1.30.1)
SDL:               2.0.8  (runtime 2.0.8)
SDL_image:         2.0.3  (runtime 2.0.3)
SDL_mixer:         2.0.2  (runtime 2.0.2)
SDL_ttf:           2.0.14 (runtime 2.0.14)

Features
========

JPEG screenshots:               yes
Lua console completion:         yes
Legacy bidirectional rendering: no
Win32 notifications back end:   yes

Current video settings
======================

Current pixel resolution: 1920x1080
Refresh rate: 59
SDL reports: 96x96 dots per inch.

Installed add-ons
=================

A_Little_Adventure : 4.1.1
A_New_Order : 1.3.0
Akladian_Music : 1.0.1
Flight_Freedom : 5.3.3 (BFW 1.14.0+)
Son_Of_The_Black_Eye_Easy_Edition : 1.2.0
The_Earths_Gut : 0.3.0
The_War_of_Terrador : 1.4.0
Two_Brothers_With_A_Third_Difficulty : 1.0.2
```
We were just playing the tutorial when a graphical glitch arose similar to the one in 4855, and then the crash happened. The only real unusual thing about our setup was playing with a heavily increased zoom level, since she couldn't see the units very well.
Logs from both times this crash happened:
[wesnoth-20210816-092021-5784.log](https://github.com/wesnoth/wesnoth/files/6995911/wesnoth-20210816-092021-5784.log)
[wesnoth-20210816-163046-13568.log](https://github.com/wesnoth/wesnoth/files/6995912/wesnoth-20210816-163046-13568.log)
Saves from around the crash:
[Wesnoth Tutorial — Part II Turn 1.gz](https://github.com/wesnoth/wesnoth/files/6995914/Wesnoth.Tutorial.Part.II.Turn.1.gz)
[Wesnoth Tutorial — Part II-Auto-Save2.gz](https://github.com/wesnoth/wesnoth/files/6995915/Wesnoth.Tutorial.Part.II-Auto-Save2.gz)
[Wesnoth Tutorial — Part II-Auto-Save3.gz](https://github.com/wesnoth/wesnoth/files/6995916/Wesnoth.Tutorial.Part.II-Auto-Save3.gz)
[Wesnoth Tutorial — Part II Turn 3.gz](https://github.com/wesnoth/wesnoth/files/6995917/Wesnoth.Tutorial.Part.II.Turn.3.gz)
[Wesnoth Tutorial — Part II-Auto-Save4.gz](https://github.com/wesnoth/wesnoth/files/6995918/Wesnoth.Tutorial.Part.II-Auto-Save4.gz)

**Comment by Pentarctagon** [comment]
What is the RAM usage reported in task manager at the time of the crash?

**Comment by Pentarctagon** [comment]
@irydacea Steps to reproduce:
* Go to the second scenario of the tutorial
* Zoom all the way in
* Scroll repeatedly over the map, particularly the water areas.

I wasn't able to reproduce the crash since the Linux builds of 1.14 are 64-bit, but I was pretty quickly able to get Wesnoth's RAM usage to 3 GBs.

@cooljeanius The workaround for you here is to switch to the Beta channel on Steam.  1.14 on Windows is only 32-bit, so it's limited to ~2 GBs RAM whereas 1.15+ is 64-bit and does not have that limit.

**Comment by Pentarctagon** [comment]
Turning off water animations might also help here, I suspect.  My guess would be that wesnoth's cache is caching the much larger, scaled up images, which then takes up a lot more RAM.  This would be a particular problem for the animated water given the number of frames it has.

**Comment by cooljeanius** [comment]
> Turning off water animations might also help here, I suspect. My guess would be that wesnoth's cache is caching the much larger, scaled up images, which then takes up a lot more RAM. This would be a particular problem for the animated water given the number of frames it has.

ok, turning of water animation does indeed help. Upgrading to 1.15+ isn't currently an option yet, though, since the main reason I wanted to get her into Wesnoth was to show her my add-on campaigns, which I haven't ported to 1.15+ yet... I'll get around to it eventually, though...

**Comment by Wedge009** [comment]
>     * Go to the second scenario of the tutorial
>     * Zoom all the way in
>     * Scroll repeatedly over the map, particularly the water areas.
> 
> I wasn't able to reproduce the crash since the Linux builds of 1.14 are 64-bit, but I was pretty quickly able to get Wesnoth's RAM usage to 3 GBs.

Also testing on Linux stable release, I was surprised I was able to get to 2.9-3.0 GiB of memory usage for the wesnoth process. Is there some sort of caching of interpolated graphics involved? I recall the first scenario for Dead Water is notorious for high CPU usage because of the high water to other terrain ratio and associated water animations, but starting from a fresh process instance I was only able to reach 1.1 GiB memory usage.

Should the title be changed to something relating to memory usage?

**Comment by Wedge009** [comment]
Testing the 32-bit release build (1.14.17) on Windows 7, I seemed to be capped at just under 2 GiB memory usage, but I suffered a lot of graphical artefacts upon scrolling and lots of corresponding error messages in the log relating to failure to create surface, out of memory, etc. However, I didn't get a crash until I attempted to exit the scenario (SDL_CreateSoftwareRenderer failed), but I think it's close enough to confirm this.

Not strictly Windows-specific, but added the label because only Windows (at least for now) has 32-bit release.

**Comment by Wedge009** [comment]
While it's not good to use excessive amounts of memory, can we consider closing this when 1.16 is released?

**Comment by Pentarctagon** [comment]
The excessive memory usage due to the image cache bug is tracked in another issue, so that sounds reasonable.

**Comment by cooljeanius** [comment]
> The excessive memory usage due to the image cache bug is tracked in another issue, so that sounds reasonable.

which other issue, to be clear?

**Comment by Pentarctagon** [comment]
#2915

**Comment by cooljeanius** [comment]
> 2915

ok thanks
